src/vs/workbench/contrib/scopeTree/browser/bookmarksView.ts

```diff
@@ -20,11 +20,249 @@ import { IBookmarksManager, BookmarkType } from 'vs/workbench/contrib/scopeTree/
+		return basename(this._resource);
+	}
+
+	public getParent(): string {
+		return dirname(this._resource).toString();
+	}
+
+	get resource(): URI {
+		return this._resource;
+	}
+}
+
+export class BookmarkHeader {
+	private _scope: BookmarkType;
+	private _visibility: boolean;
+
+	constructor(scope: BookmarkType) {
+		this._scope = scope;
+		this._visibility = true;
+	}
+
+	get scope(): BookmarkType {
+		return this._scope;
+	}
+
+	get visibility(): boolean {
+		return this._visibility;
+	}
+
+	set visibility(show: boolean) {
+		this._visibility = show;
+	}
+}
+
+class BookmarkDelegate implements IListVirtualDelegate<Bookmark | BookmarkHeader> {
+	static readonly ITEM_HEIGHT = 22;
+
+	getHeight(element: Bookmark | BookmarkHeader): number {
+		return BookmarkDelegate.ITEM_HEIGHT;
+	}
+
+	getTemplateId(element: Bookmark | BookmarkHeader): string {
+		if (element instanceof Bookmark) {
+			return BookmarkRenderer.ID;
+		}
+
+		return BookmarkHeaderRenderer.ID;
+	}
+}
+
+interface IBookmarkTemplateData {
+	bookmarkContainer: HTMLElement;
+	label: IResourceLabel;
+	elementDisposable: IDisposable;
+}
+
+interface IBookmarkHeaderTemplateData {
+	headerContainer: HTMLElement;
+	elementDisposable: IDisposable;
+}
+
+class BookmarkElementIconRenderer implements IDisposable {
+	private _focusIcon!: HTMLElement;
+
+	constructor(container: HTMLElement,
+		private readonly stat: URI,
+		@IExplorerService private readonly explorerService: IExplorerService) {
+		this.renderFocusIcon(container);
+		this.addListeners(container);
+	}
+
+	get focusIcon(): HTMLElement {
+		return this._focusIcon;
+	}
+
+	private addListeners(container: HTMLElement): void {
+		container.addEventListener('mouseover', () => {
+			this._focusIcon.style.visibility = 'visible';
+		});
+		container.addEventListener('mouseout', () => {
+			this._focusIcon.style.visibility = 'hidden';
+		});
+		container.addEventListener('dblclick', async () => {
+			await this.explorerService.select(this.stat, true);	// Should also expand directory
+		});
+		this._focusIcon.addEventListener('click', () => {
+			this.explorerService.setRoot(this.stat);
+		});
+	}
+
+	private renderFocusIcon(container: HTMLElement): void {
+		this._focusIcon = document.createElement('img');
+		this._focusIcon.className = 'scope-tree-focus-icon-near-bookmark';
+		container.insertBefore(this._focusIcon, container.firstChild);
+	}
+
+	dispose(): void {
+		this._focusIcon.remove();
+	}
+}
+
+class BookmarkElementHeaderRenderer implements IDisposable {
+	private headerContainer: HTMLElement;
+
+	constructor(headerContainer: HTMLElement) {
+		this.headerContainer = headerContainer;
+	}
+
+	dispose(): void {
+		this.headerContainer.remove();
+	}
+}
+
+class BookmarkRenderer implements ITreeRenderer<Bookmark, FuzzyScore, IBookmarkTemplateData> {
+	static readonly ID = 'BookmarkRenderer';
+
+	constructor(
+		private labels: ResourceLabels,
+		private readonly explorerService: IExplorerService
+	) {
+		// noop
+	}
```

Comment:
As in a previous PR:

```suggestion
	constructor(
		private labels: ResourceLabels,
		private readonly explorerService: IExplorerService
	) {}
```